WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/AuthorSelectionAdapter.kt

```diff
@@ -0,0 +1,137 @@
+import android.support.v4.content.ContextCompat
+import android.support.v7.widget.AppCompatImageView
+import android.support.v7.widget.AppCompatTextView
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.SpinnerAdapter
+import androidx.annotation.CallSuper
+import org.wordpress.android.R
+import org.wordpress.android.WordPress
+import org.wordpress.android.ui.posts.PostListMainViewModel.AuthorFilterListItemUIState
+import org.wordpress.android.ui.utils.UiHelpers
+import org.wordpress.android.util.GravatarUtils
+import org.wordpress.android.util.image.ImageManager
+import java.lang.ref.WeakReference
+import javax.inject.Inject
+
+class AuthorSelectionAdapter(context: Context) : SpinnerAdapter {
+    @Inject lateinit var imageManager: ImageManager
+    @Inject lateinit var uiHelpers: UiHelpers
+
+    private var observers: MutableList<WeakReference<DataSetObserver>> = mutableListOf()
+    private val items = mutableListOf<AuthorFilterListItemUIState>()
+
+    init {
+        (context.applicationContext as WordPress).component().inject(this)
+    }
+
+    override fun getCount(): Int = items.count()
+
+    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
+        var view: View? = convertView
+        val holder: DropdownViewHolder
+
+        if (view == null) {
+            val inflater = LayoutInflater.from(parent.context)
+            view = inflater.inflate(R.layout.post_list_author_selection_dropdown, parent, false)
+            holder = DropdownViewHolder(view)
+            view.tag = holder
+        } else {
+            holder = view.tag as DropdownViewHolder
+        }
+
+        holder.bind(items[position], imageManager, uiHelpers)
+
+        return view!!
+    }
+
+    override fun getItem(position: Int): AuthorFilterListItemUIState = items[position]
+
+    override fun getItemId(position: Int): Long = position.toLong()
+
+    override fun getItemViewType(position: Int): Int = 0
+
+    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+        var view: View? = convertView
+        val holder: NormalViewHolder
+
+        if (view == null) {
+            val inflater = LayoutInflater.from(parent.context)
+            view = inflater.inflate(R.layout.post_list_author_selection, parent, false)
+            holder = NormalViewHolder(view)
+            view.tag = holder
+        } else {
+            holder = view.tag as NormalViewHolder
+        }
+
+        holder.bind(items[position], imageManager, uiHelpers)
+
+        return view!!
+    }
+
+    override fun getViewTypeCount(): Int = 1
+    override fun hasStableIds(): Boolean = false
+    override fun isEmpty(): Boolean = false
+
+    fun notifyDataSetChanged() {
+        for (observer in observers) {
+            observer.get()?.onChanged()
+        }
+    }
+
+    override fun registerDataSetObserver(observer: DataSetObserver) {
```

Comment:
I think we can improve this whole method with something like this:

```
val weakObserver = WeakReference(observer)
if (!observers.contains(weakObserver)) {
    observers.add(weakObserver)
}
```

Actually can't we use a `MutableSet` for the observers instead and just avoid this whole thing?

Reply:
Sure

Reply:
Actually looking at the code @oguzkocer we can't rely on contains given that neither `WeakReference` or its super class `Reference` override the `equals` function. Which is  As such we need to call `get` and do that comparison ourselves.

Do `MutableSet`s store things as `WeakReference`s? Or something similar? Or am I missing some other benefit?

Reply:
Interesting, good find. I guess in the context of `Reference` what `equals` means is not clear, so it's not implemented for it. I _think_ we can still use a set if we wanted to, but it's not really necessary, feel free to resolve this conversation.

For reference, I believe this code will work, but I haven't tested it:

```
if (!observers.mapNotNull { it.get() }.contains(observer)) {
    observers.add(WeakReference(observer))
}
```

Reply:
Implemented that code change in 79fd384edcc91c147f76b1843cc78ade7e9caa6c